compiler/modules/types/common.bal

```diff
@@ -52,19 +56,29 @@ function memoSubtypeIsEmpty(Context cx, BddMemoTable memoTable, BddIsEmptyPredic
     int initStackDepth = cx.memoStack.length();
     cx.memoStack.push(m);
     boolean isEmpty = isEmptyPredicate(cx, b);
+    boolean cyclic = isEmpty && (m.empty == "loop");
     if !isEmpty || initStackDepth == 0 {
         foreach int i in initStackDepth + 1 ..< cx.memoStack.length() {
             MemoEmpty memoEmpty = cx.memoStack[i].empty;
-            if memoEmpty == "provisional" {
+            if memoEmpty is "provisional"|"loop"|"cyclic" {
                 cx.memoStack[i].empty = isEmpty ? isEmpty : ();
             }
         }
         cx.memoStack.setLength(initStackDepth);
         m.empty = isEmpty;
     }
+    if cyclic {
+        m.empty = "cyclic";    
+    }
     return isEmpty;
 }
 
+function memoSubtypeIsCyclic(Context cx, BddMemoTable memoTable, BddIsEmptyPredicate isEmptyPredicate, Bdd b) returns boolean {
+    // This assume we have already checked (and confirmed) type to be empty
```

Comment:
It only assumes that we have tested (and so memoized) whether the type is empty.

Reply:
fixed